gattc_ninafw.go

```diff
@@ -0,0 +1,318 @@
+		}
+	default:
+		for _, s := range foundServices {
+			services = append(services, s)
+		}
+	}
+
+	return services, nil
+}
+
+// DeviceCharacteristic is a BLE characteristic on a connected peripheral
+// device.
+type DeviceCharacteristic struct {
+	uuid UUID
+
+	service     *DeviceService
+	permissions CharacteristicPermissions
+	handle      uint16
+	properties  uint8
+	callback    func(buf []byte)
+}
+
+// UUID returns the UUID for this DeviceCharacteristic.
+func (c *DeviceCharacteristic) UUID() UUID {
+	return c.uuid
+}
+
+// DiscoverCharacteristics discovers characteristics in this service. Pass a
+// list of characteristic UUIDs you are interested in to this function. Either a
+// list of all requested services is returned, or if some services could not be
+// discovered an error is returned. If there is no error, the characteristics
+// slice has the same length as the UUID slice with characteristics in the same
+// order in the slice as in the requested UUID list.
+//
+// Passing a nil slice of UUIDs will return a complete
+// list of characteristics.
+func (s *DeviceService) DiscoverCharacteristics(uuids []UUID) ([]DeviceCharacteristic, error) {
+	if _debug {
+		println("DiscoverCharacteristics")
+	}
+
+	characteristics := make([]DeviceCharacteristic, 0, maxDefaultCharacteristicsToDiscover)
+	foundCharacteristics := make(map[UUID]DeviceCharacteristic)
+
+	startHandle := s.startHandle
+	endHandle := s.endHandle
+	for startHandle < endHandle {
+		err := s.device.adapter.att.readByTypeReq(s.device.handle, startHandle, endHandle, gattCharacteristicUUID)
+		switch {
+		case err == ErrATTOp &&
+			s.device.adapter.att.lastErrorOpcode == attOpReadByTypeReq &&
+			s.device.adapter.att.lastErrorCode == attErrorAttrNotFound:
+
+			// no characteristics found
+			break
+		case err != nil:
+			return nil, err
+		}
+
+		if _debug {
+			println("found s.device.adapter.att.characteristics", len(s.device.adapter.att.characteristics))
+		}
+
+		if len(s.device.adapter.att.characteristics) == 0 {
+			break
+		}
+
+		for _, rawCharacteristic := range s.device.adapter.att.characteristics {
+			if len(uuids) == 0 || rawCharacteristic.uuid.IsIn(uuids) {
+				permissions := CharacteristicPermissions(0)
+				if rawCharacteristic.properties&charPropertyBroadcast != 0 {
+					permissions |= CharacteristicBroadcastPermission
+				}
+				if rawCharacteristic.properties&charPropertyRead != 0 {
+					permissions |= CharacteristicReadPermission
+				}
+				if rawCharacteristic.properties&charPropertyWriteWithoutResponse != 0 {
+					permissions |= CharacteristicWriteWithoutResponsePermission
+				}
+				if rawCharacteristic.properties&charPropertyWrite != 0 {
+					permissions |= CharacteristicWritePermission
+				}
+				if rawCharacteristic.properties&charPropertyNotify != 0 {
+					permissions |= CharacteristicNotifyPermission
+				}
+				if rawCharacteristic.properties&charPropertyIndicate != 0 {
+					permissions |= CharacteristicIndicatePermission
+				}
```

Comment:
It looks to me that you could use `rawCharacteristic.properties` directly because all the values are exactly the same as the `CharacteristicPermissions` enum.

Reply:
You are correct. Removing.